crates/bevy_ecs/src/storage/blob_array.rs

```diff
@@ -0,0 +1,455 @@
+                // * 0 <= `i` < `len`, so `i * size` must be in bounds for the allocation.
+                // * `size` is a multiple of the erased type's alignment,
+                //   so adding a multiple of `size` will preserve alignment.
+                // * The item is left unreachable so it can be safely promoted to an `OwningPtr`.
+                let item = unsafe { self.get_ptr_mut().byte_add(i * size).promote() };
+                // SAFETY: `item` was obtained from this `BlobArray`, so its underlying type must match `drop`.
+                unsafe { drop(item) };
+            }
+            self.drop = Some(drop);
+        }
+    }
+
+    /// Because this method needs parameters, it can't be the implementation of the `Drop` trait.
+    /// The owner of this [`BlobArray`] must call this method with the correct information.
+    ///
+    /// # Safety
+    /// - `cap` and `len` are indeed the capacity and length of this [`BlobArray`]
+    /// - This [`BlobArray`] mustn't be used after calling this method.
+    pub unsafe fn drop(&mut self, cap: usize, len: usize) {
+        if cap != 0 {
+            self.clear_elements(len);
+            if !self.is_zst() {
+                let layout =
+                    array_layout(&self.item_layout, cap).expect("array layout should be valid");
+                std::alloc::dealloc(self.data.as_ptr().cast(), layout);
+            }
+        }
+    }
+
+    /// Drops the last element in this [`BlobArray`].
+    ///
+    /// # Safety
+    // - `last_element_index` must correspond to the last element in the array.
+    // - After this method is called, the last element must not be used
+    // unless [`Self::initialize_unchecked`] is called to set the value of the last element.
+    pub unsafe fn drop_last_element(&mut self, last_element_index: usize) {
+        if let Some(drop) = self.drop {
+            // We set `self.drop` to `None` before dropping elements for unwind safety. This ensures we don't
+            // accidentally drop elements twice in the event of a drop impl panicking.
+            self.drop = None;
+            // SAFETY:
+            let item = self.get_unchecked_mut(last_element_index).promote();
+            // SAFETY:
+            unsafe { drop(item) };
+            self.drop = Some(drop);
+        }
+    }
+
+    /// Allocate a block of memory for the array. This should be used to initialize the array, do not use this
+    /// method if there are already elements stored in the array - use [`Self::realloc`] instead.
+    pub(super) fn alloc(&mut self, capacity: NonZeroUsize) {
+        if !self.is_zst() {
+            let new_layout = array_layout(&self.item_layout, capacity.get())
+                .expect("array layout should be valid");
+            let new_data
+            // SAFETY:
+            // - layout has non-zero size because capacity > 0, and the blob isn't ZST (`self.is_zst` == false)
+            = unsafe {std::alloc::alloc(new_layout)};
```

Comment:
This formatting looks pretty weird. I think the position of the comment confuses `rustfmt`, you should try moving it over the `let new_data`.